<h1> HOMEWORK 1 </h1>
<h3> Background Information on data </h3>
<p>The dataset was dervived from an animal shelter in Austin, Texas. The dataset contains information regarding the animals that was brought to the shelter and the animals final outcome whether it was succesfully adopted/returned to owner or if animal died at the shelter. The final outcome of 1 indicates death while 0 indicates that animal was adopted/returned

<h2>Part 1 Data Prepration and Cleaning</h2>
    <p>Objective</p>
    <ol><li>Import all required libraries</li>
    <li>Find initial dimension of data</li>
        <li>Print the first and the last 5 rows.</li>
        <li>Convert the features to their appropriate data types </li>
        <li>Drop duplicate rows and columns, if any.</li>
        <li>Save your updated/cleaned data frame to a new csv file.</li>

<h4>1.Import libraries</h4>
<p> The most important library used is the <b>pandas</b>, this enables the csv to be converted to a pandas datatype and allow data manipulation. <b>Numpy</b> library adds array like functionalities to python for scientific calculations. The visulaization is carried out using <b>Matplotlib</b> and <b>Seaborn</b>.

In [52]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd

In [53]:
# This option enables auto complete in notebook
%config IPCompleter.greedy=True

In [54]:
# Load csv to pandas dataframe
df = pd.read_csv('AnimalWelfareRisk-19200494.csv')
# Enables viewing of all columns when using the head and tail command
pd.options.display.max_columns = None
pd.options.display.max_rows = None


# Show full column with out truncation
pd.set_option('display.max_colwidth', -1)

<h4>2.Get initial dimension of dataframe</h4>

In [55]:
df.shape

(1000, 22)

<blockquote>The dataframe initially has dimension of 1000 x 22 which means there are 22 columns of attributes and 1000 rows of data</blockquote>

<h4>3.Getting an overview of data </h4>
<p>by checking first 5 and last 5 data by using head() and tail() functions</p>

In [56]:
df.head()

Animal ID Name_Intake         DateTime_Intake        MonthYear_Intake  \
0  A737831   NaN         11/05/2016 12:12:00 PM  11/05/2016 12:12:00 PM   
1  A637559   Blue        09/28/2016 11:55:00 AM  09/28/2016 11:55:00 AM   
2  A704521   *Goldie     06/05/2015 05:18:00 PM  06/05/2015 05:18:00 PM   
3  A703884   NaN         05/29/2015 11:45:00 PM  05/29/2015 11:45:00 PM   
4  A718401   NaN         12/29/2015 12:58:00 PM  12/29/2015 12:58:00 PM   

                        Found Location    Intake Type Intake Condition  \
0  3604 Alta Ct in Austin (TX)          Stray          Normal            
1  Austin (TX)                          Public Assist  Normal            
2  5915 Old Lockhart Rd in Travis (TX)  Stray          Nursing           
3  7102 Darcus Cv in Austin (TX)        Wildlife       Normal            
4  1201 Rutland Dr in Austin (TX)       Stray          Normal            

  Animal Type_Intake Sex upon Intake Age upon Intake            Breed_Intake  \
0  Cat                Neutered Male   5 months        Domestic Shorthair Mix   
1  Dog                Intact Male     5 years         Pit Bull                 
2  Cat                Intact Female   1 year          Domestic Shorthair Mix   
3  Other              Unknown         4 weeks         Bat                      
4  Dog                Intact Male     1 month         Labrador Retriever Mix   

         Color_Intake Name_Outcome        DateTime_Outcome  \
0  White/Orange Tabby  NaN          11/30/2016 02:23:00 PM   
1  Blue/White          Blue         10/08/2016 11:21:00 AM   
2  Orange Tabby        *Goldie      06/24/2015 04:41:00 PM   
3  Black/Brown         NaN          05/31/2015 08:44:00 AM   
4  Black/White         NaN          01/02/2016 12:09:00 PM   

        MonthYear_Outcome Date of Birth Animal Type_Outcome Sex upon Outcome  \
0  11/30/2016 02:23:00 PM  06/05/2016    Cat                 Neutered Male     
1  10/08/2016 11:21:00 AM  09/19/2011    Dog                 Intact Male       
2  06/24/2015 04:41:00 PM  06/05/2014    Cat                 Spayed Female     
3  05/31/2015 08:44:00 AM  04/29/2015    Other               Unknown           
4  01/02/2016 12:09:00 PM  10/29/2015    Dog                 Intact Male       

  Age upon Outcome           Breed_Outcome       Color_Outcome  binary_outcome  
0  5 months         Domestic Shorthair Mix  White/Orange Tabby  0.0             
1  5 years          Pit Bull                Blue/White          0.0             
2  1 year           Domestic Shorthair Mix  Orange Tabby        0.0             
3  4 weeks          Bat                     Black/Brown         1.0             
4  2 months         Labrador Retriever Mix  Black/White         0.0

In [57]:
df.tail()

Animal ID Name_Intake         DateTime_Intake        MonthYear_Intake  \
995  A809500   NaN         11/25/2019 03:33:00 PM  11/25/2019 03:33:00 PM   
996  A709273   *Lokken     08/06/2015 04:45:00 PM  08/06/2015 04:45:00 PM   
997  A750040   Olivia      10/11/2017 05:50:00 PM  10/11/2017 05:50:00 PM   
998  A669789   Hawkeye     12/27/2013 01:14:00 PM  12/27/2013 01:14:00 PM   
999  A669967   Nikko       12/31/2013 10:49:00 AM  12/31/2013 10:49:00 AM   

                      Found Location      Intake Type Intake Condition  \
995  Volente Beach in Travis (TX)     Stray            Normal            
996  Austin (TX)                      Owner Surrender  Normal            
997  Austin (TX)                      Owner Surrender  Normal            
998  Perth Pass in Pflugerville (TX)  Stray            Normal            
999  Austin (TX)                      Owner Surrender  Normal            

    Animal Type_Intake Sex upon Intake Age upon Intake  \
995  Dog                Intact Female   8 months         
996  Cat                Intact Male     1 year           
997  Dog                Spayed Female   7 years          
998  Dog                Intact Male     2 years          
999  Dog                Intact Male     1 month          

                Breed_Intake Color_Intake Name_Outcome  \
995  Great Dane               Black/White  NaN           
996  Domestic Shorthair Mix   Black        *Lokken       
997  Labrador Retriever Mix   Brown        Olivia        
998  Labrador Retriever Mix   Yellow       Hawkeye       
999  Chihuahua Shorthair Mix  Brown/White  Nikko         

           DateTime_Outcome       MonthYear_Outcome Date of Birth  \
995  12/04/2019 06:20:00 PM  12/04/2019 06:20:00 PM  03/25/2019     
996  08/15/2015 03:23:00 PM  08/15/2015 03:23:00 PM  04/06/2014     
997  05/02/2018 08:15:00 AM  05/02/2018 08:15:00 AM  12/29/2009     
998  12/30/2013 12:22:00 PM  12/30/2013 12:22:00 PM  12/28/2011     
999  12/31/2013 05:15:00 PM  12/31/2013 05:15:00 PM  10/31/2013     

    Animal Type_Outcome Sex upon Outcome Age upon Outcome  \
995  Dog                 Spayed Female    8 months          
996  Cat                 Intact Male      1 year            
997  Dog                 Spayed Female    8 years           
998  Dog                 Neutered Male    2 years           
999  Dog                 Intact Male      1 month           

               Breed_Outcome Color_Outcome  binary_outcome  
995  Great Dane               Black/White   0.0             
996  Domestic Shorthair Mix   Black         0.0             
997  Labrador Retriever Mix   Brown         0.0             
998  Labrador Retriever Mix   Yellow        0.0             
999  Chihuahua Shorthair Mix  Brown/White   0.0

<h4>4. Checking datatypes of all attributes</h4>
Almost all the data is of  object type with only final output being a float type
need more exploration to be done to check if all data is to be kept as object or any other types

In [58]:
df.dtypes

Animal ID              object 
Name_Intake            object 
DateTime_Intake        object 
MonthYear_Intake       object 
Found Location         object 
Intake Type            object 
Intake Condition       object 
Animal Type_Intake     object 
Sex upon Intake        object 
Age upon Intake        object 
Breed_Intake           object 
Color_Intake           object 
Name_Outcome           object 
DateTime_Outcome       object 
MonthYear_Outcome      object 
Date of Birth          object 
Animal Type_Outcome    object 
Sex upon Outcome       object 
Age upon Outcome       object 
Breed_Outcome          object 
Color_Outcome          object 
binary_outcome         float64
dtype: object

<p> Since most of the datatypes is of type object we will not get much information by describing data</p> The data types of each attribute will be analyzed and converted in sections below


In [59]:
df.describe().T
# Such results not usable due to data types

count   mean       std  min  25%  50%  75%  max
binary_outcome  1000.0  0.088  0.283437  0.0  0.0  0.0  0.0  1.0

<h4>5. Finding duplicate columns </h4>
A function called <b>find_duplicates(dataframe)</b> is used to find all duplicate columns in dataframe (which doesn't have null values).
<br>The function accepts a dataframe and goes through each column while comparing the contents with other columns<br>If a column is found to have same content with the compared column. The column ids are added as pair to a duplicate list and returned in the end

In [60]:
def find_duplicates(dataframe):
    x = dataframe.columns
    y = dataframe.columns
    row_size = dataframe.shape[0]
    duplicate_pairs = []
    for index1 in range (len(x)):
        for index2 in range (index1+1,len(y)):
            if sum(df[x[index1]]==df[y[index2]])==row_size:
                duplicate_pairs.append([x[index1],y[index2]])
    return duplicate_pairs

In [61]:
duplicate_pairs=find_duplicates(df)
print("The duplicate pairs are :\n")
for x,y in  duplicate_pairs:
    print (" {} --> duplicate of column --> {}\n".format(x,y))

The duplicate pairs are :

 DateTime_Intake --> duplicate of column --> MonthYear_Intake

 Animal Type_Intake --> duplicate of column --> Animal Type_Outcome

 Breed_Intake --> duplicate of column --> Breed_Outcome

 Color_Intake --> duplicate of column --> Color_Outcome

 DateTime_Outcome --> duplicate of column --> MonthYear_Outcome



<blockquote>Observations</blockquote>
By executing the function to find duplicates its observed that there are 5 pairs of duplicate columns in dataframe. An important thing to note is that the function only checks for columns which have all values. Therefor more investigation is to be carried out to see if there are columns which have duplicates and null values

<h4> Finding out all the null values in attributes</h4>

In [62]:
df[df.columns].isnull().sum()

Animal ID              0  
Name_Intake            352
DateTime_Intake        0  
MonthYear_Intake       0  
Found Location         0  
Intake Type            0  
Intake Condition       0  
Animal Type_Intake     0  
Sex upon Intake        0  
Age upon Intake        0  
Breed_Intake           0  
Color_Intake           0  
Name_Outcome           352
DateTime_Outcome       0  
MonthYear_Outcome      0  
Date of Birth          0  
Animal Type_Outcome    0  
Sex upon Outcome       0  
Age upon Outcome       1  
Breed_Outcome          0  
Color_Outcome          0  
binary_outcome         0  
dtype: int64

<blockquote>Observation</blockquote>
It seems that 'Name_Intake' and 'Name_outcome' are the only columns where there seems to be missing information. Around 35.2 % of data is missiong. Another interesting observation is that the exact amount of data is also missing. This means thses two might be duplicates of each other and since the find_duplicate function only checks functions which have full information. We will have to check these two individually

In [63]:
sum(df['Name_Intake']==df['Name_Outcome'])

648

<blockquote> Observation </blockquote>Here upon checking columns 'Name_intake' and 'Name_Outcome' both seems to have same content for all 648 available entries
the rest 352 are null , so we can say by sure most of elemnts are duplicates of each column

<h3>6. Dropping columns </h3>
<p>Most real life datasets contains some form of errors , unnecessary attributes or missing information. In such cases the dataset must be cleaned before it is used inside a model for training. The same case is applicable in this scenario as well. In the given dataset there seems to be duplicate columns, missing data and redundant information. Therefor it is important that we drop these unnecessary columns. Detailed explaination for droping them will be provided below<p>

<p> The most obvious columns that will be dropped are the duplicate ones as holding on to redundant data is of no value. so columns <br>
<ul><li>MonthYear_Intake</li>
    <li>Animal Type_Outcome</li>
    <li>Breed_Outcome</li>
    <li>Color_Outcome</li>
    <li>MonthYear_Outcome</li>
<br>Will be dropped from dataset.

In [64]:
for x,y in duplicate_pairs:
    df = df.drop(y,axis=1)

After droping the colums of new dataframe is checked using df.head(1). And it shows that all duplicates where dropped.

In [65]:
df.head(1)

Animal ID Name_Intake         DateTime_Intake               Found Location  \
0  A737831   NaN         11/05/2016 12:12:00 PM  3604 Alta Ct in Austin (TX)   

  Intake Type Intake Condition Animal Type_Intake Sex upon Intake  \
0  Stray       Normal           Cat                Neutered Male    

  Age upon Intake            Breed_Intake        Color_Intake Name_Outcome  \
0  5 months        Domestic Shorthair Mix  White/Orange Tabby  NaN           

         DateTime_Outcome Date of Birth Sex upon Outcome Age upon Outcome  \
0  11/30/2016 02:23:00 PM  06/05/2016    Neutered Male    5 months          

   binary_outcome  
0  0.0

<p>Next attibutes we look at are the "Name_Intake" and "Name_Outcome" we look at it because of <b>three</b> reasons<br>
<ul> <li>Almost 35.2% of the data is missing in both the attributes.</li>
        <li> Of the available 64.8% data both seems to share exact data. Therefor it seems both are duplicate of each others. Imputation is not possible in this data as these are names and we cannot use statistical inferance to impute data.</li>
    <li>Since these data are just names the attibutes hold no meaning to overall analysis. </li></ul>
<p> Considering all these factors its best to remove both the attributes. 

In [66]:
df = df.drop(["Name_Intake","Name_Outcome"],axis=1)

<p> Another attribute which can be removed is the attribute "Animal ID" because it unlike noraml index the attribute is too long to be refference using .loc() and instead can be replaced with normal indexing from 1 to 1000</p>

In [67]:
df = df.drop(["Animal ID"],axis=1)

<p> The column date of birth can be used to find age of an animal, However, There is already a column called 'Age upon outcome' which will the same result. Therfore column 'date of birth' can be removed as same data can be derived from other columns as well</p>

In [68]:
df.drop(columns="Date of Birth",inplace = True , axis=1)

Here I am adding an index to dataframe which is from 0 - 999 so that we can use slice operators like iloc later 

In [69]:
index = [x for x in range (1000)]
df["index"] = index
df["index"] = df.index

In [70]:
df.head(1)

DateTime_Intake               Found Location Intake Type  \
0  11/05/2016 12:12:00 PM  3604 Alta Ct in Austin (TX)  Stray        

  Intake Condition Animal Type_Intake Sex upon Intake Age upon Intake  \
0  Normal           Cat                Neutered Male   5 months         

             Breed_Intake        Color_Intake        DateTime_Outcome  \
0  Domestic Shorthair Mix  White/Orange Tabby  11/30/2016 02:23:00 PM   

  Sex upon Outcome Age upon Outcome  binary_outcome  index  
0  Neutered Male    5 months         0.0             0

<h3>7. Data Types</h3>

<p> Since we have cleaned up most of the data of null values and redundancies. The focus should be for now shifted towards the datatypes of attributes. By looking at the dataframe inspected in section 4 (above) we found out that almost all data are of type object and only the final outcome is of type float. For a proper exploratory analysis to be done the datatypes for attibutes must be a proper one. The most common of the two are categorical and continious type.</p>


<p>First thing to check are the cardinalities of the attributes. Any attributes with cardiality of 1 must be investigated as it will not be able to provide usefull information.</p> <br> Therefor we investigate the cardinalities of each attribute 

In [71]:
def get_cardinality(dataframe):
    list_colmn = []
    list_data = []
    print("The cardinalities of all the columns are\n")
    for x in dataframe.columns:
        list_colmn.append(x)
        list_data.append(dataframe[x].nunique())
    data = {'Attributes':list_colmn, 'Cardinalities':list_data}
    display(pd.DataFrame(data))

In [72]:
get_cardinality(df)

The cardinalities of all the columns are



Attributes  Cardinalities
0   DateTime_Intake     996          
1   Found Location      759          
2   Intake Type         5            
3   Intake Condition    8            
4   Animal Type_Intake  4            
5   Sex upon Intake     5            
6   Age upon Intake     42           
7   Breed_Intake        219          
8   Color_Intake        120          
9   DateTime_Outcome    1000         
10  Sex upon Outcome    5            
11  Age upon Outcome    42           
12  binary_outcome      2            
13  index               1000


<blockquote>Observation<br><p> From the cardinalities printed we can see that there are no attributes with a cardinality of just 1; Another observation is the fact that there are some attributes with cardinalities less than the number of instances. Such attributes are good candidates for Categorical datatypes.</p></blockquote>

<p> First we will try to find all the low cardinality attributes which have cardinalities less than 10 and analyse them first</p>

In [73]:
def filter_elements_by_cardinality(dataframe,cardinality):
    
    low_cardinality_list = dataframe.columns[dataframe[dataframe.columns].nunique() < cardinality ]
    low_cardinality_list_info =dataframe[low_cardinality_list].nunique()
    data ={"Cardinality":low_cardinality_list_info}
    display(pd.DataFrame(data))
    return(low_cardinality_list)

In [74]:
low_cardinality_columns = filter_elements_by_cardinality(df,10)

Cardinality
Intake Type         5          
Intake Condition    8          
Animal Type_Intake  4          
Sex upon Intake     5          
Sex upon Outcome    5          
binary_outcome      2


<blockquote>Observation<br><p> By filtering out first attributes with Cardinality less than 10. We discovered 6 attributes that fit the category.</p></blockquote>

In [75]:
def list_unique_elements(df,columns):
    list = []
    list_index = []
    for x in columns:
        list.append(df[x].unique())
        list_index.append(x)
    data = {"Atributes":list_index,"Unique elements":list}
    display(pd.DataFrame(data))

In [76]:
list_unique_elements(df,low_cardinality_columns)

Atributes  \
0  Intake Type          
1  Intake Condition     
2  Animal Type_Intake   
3  Sex upon Intake      
4  Sex upon Outcome     
5  binary_outcome       

                                                         Unique elements  
0  [Stray, Public Assist, Wildlife, Owner Surrender, Euthanasia Request]  
1  [Normal, Nursing, Injured, Sick, Other, Aged, Feral, Medical]          
2  [Cat, Dog, Other, Bird]                                                
3  [Neutered Male, Intact Male, Intact Female, Unknown, Spayed Female]    
4  [Neutered Male, Intact Male, Spayed Female, Unknown, Intact Female]    
5  [0.0, 1.0]

<ul><li>For the given attributes the cardinality of the attributes are much less than number of instances in dataset</li>
    <li>The observed data is in no way continious</li>
    <li> Fields like 'binary_outcome' and all the above attributes have fixed categories</li>
    <li> No meaning full data can be extracted from mean of say 'intake condition' and other attributes hence another reason to consider categorical type</li>
    <blockquote> Considering all the above factors its concluded that all the above attributes are to be converted to categorical type</blockquote>

In [77]:
for x in low_cardinality_columns:
    df[x]=df[x].astype('category')

<h3>Data Transformation </h3>
Though this is being asked to do in later section i feel this must be done here itself<br>
Some data such as 'Date Time' , 'Found location' , 'Age upon intake' , 'Date Time_outcome' , 'age upon outcome' are of type objects (String) and must be processed to extract usefull information and remove unnecessary ones so that we can use them for exploratory analysis

First attribute which will be processed to get usefull information will be 'DateTime_intake' to extract wheeter the animal came in the morning or at night. The intuition behind this is to find if there is a trend in animal abandoned at morning more or at night. This data if usefull can be used to identify peak times and allocate more resource at that time in shelter.

In [78]:
def get_AM_PM(df):
    list =[]
    for x in range (1000):
        z = df['DateTime_Intake'][x].split()
        list.append(z[2])
    return list

In [79]:
x = get_AM_PM(df)
df['intake_am_pm'] = x
# Convert it to category as well
df['intake_am_pm'] = df['intake_am_pm'].astype('category');

Next we extract just the date. Since the date follows a definite pattern regular expression is used to extract the data. The data is then converted to date time format so that we can do exploration on it like what year most animals where brought and such

In [80]:
extr = df['DateTime_Intake'].str.extract(r'(\d{2}([/])\d{1,2}([/])\d{1,4})', expand=False)
df['intake_time'] = extr[0]
df['intake_time'] = pd.to_datetime(df['intake_time'])

Since we have extracted usefull information from the 'DateTime_Intake' attibutes we can now drop it

In [81]:
df.drop(['DateTime_Intake'],axis = 1, inplace=True)

The next attribute to get feature extraction is 'Found Location' we take try to extract from where most of the animals are abandoned. The reason for this process is because if we check the attribute we initially find that the cardinality of the data is around 759 out of 1000. Which indicates that the data might to too unique to be used to find any relations. Observing the data one by one its found that the data is to specific ie almost all data contains data upto the street in which the animal is discovered. Such high granularity is not required instead we extract the major region where each animal is taken from which gives better information

In [82]:
for x in range(1000):
    z = df['Found Location'][x].split(' in')
    if len(z) == 1:
        df['Found Location'][x] = z[0].strip()
    else:
        df['Found Location'][x] = z[1].strip()

/opt/anaconda3/envs/data_analytics/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/opt/anaconda3/envs/data_analytics/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


After extracting all the general location. I was able to remove the high granularity of 'Found location' data from 756 unique values to 16 unique locations which provides more information to the attribute

In [83]:
print("The number of unique locations is: " + str(df['Found Location'].nunique()))

The number of unique locations is: 16


In [84]:
k = df['Found Location'].value_counts()

All the major location and animals picked from the place

In [85]:
display(pd.DataFrame(k))

Found Location
Austin (TX)           828           
Travis (TX)           82            
Manor (TX)            22            
Pflugerville (TX)     22            
Del Valle (TX)        16            
Outside Jurisdiction  11            
Lago Vista (TX)       4             
Leander (TX)          4             
Jonestown (TX)        3             
Bee Cave (TX)         2             
Bastrop (TX)          1             
Cedar Park (TX)       1             
Mustang Ridge (TX)    1             
Lakeway (TX)          1             
Sunset Valley (TX)    1             
Williamson (TX)       1

Now that we have categorized the found location data its time for us to convert the datatype from object type to categorical

In [86]:
df['Found Location']=df['Found Location'].astype('category');

In [87]:
df.dtypes

Found Location        category      
Intake Type           category      
Intake Condition      category      
Animal Type_Intake    category      
Sex upon Intake       category      
Age upon Intake       object        
Breed_Intake          object        
Color_Intake          object        
DateTime_Outcome      object        
Sex upon Outcome      category      
Age upon Outcome      object        
binary_outcome        category      
index                 int64         
intake_am_pm          category      
intake_time           datetime64[ns]
dtype: object

In [88]:
df.head(5)

Found Location    Intake Type Intake Condition Animal Type_Intake  \
0  Austin (TX)    Stray          Normal           Cat                 
1  Austin (TX)    Public Assist  Normal           Dog                 
2  Travis (TX)    Stray          Nursing          Cat                 
3  Austin (TX)    Wildlife       Normal           Other               
4  Austin (TX)    Stray          Normal           Dog                 

  Sex upon Intake Age upon Intake            Breed_Intake        Color_Intake  \
0  Neutered Male   5 months        Domestic Shorthair Mix  White/Orange Tabby   
1  Intact Male     5 years         Pit Bull                Blue/White           
2  Intact Female   1 year          Domestic Shorthair Mix  Orange Tabby         
3  Unknown         4 weeks         Bat                     Black/Brown          
4  Intact Male     1 month         Labrador Retriever Mix  Black/White          

         DateTime_Outcome Sex upon Outcome Age upon Outcome binary_outcome  \
0  11/30/2016 02:23:00 PM  Neutered Male    5 months         0.0             
1  10/08/2016 11:21:00 AM  Intact Male      5 years          0.0             
2  06/24/2015 04:41:00 PM  Spayed Female    1 year           0.0             
3  05/31/2015 08:44:00 AM  Unknown          4 weeks          1.0             
4  01/02/2016 12:09:00 PM  Intact Male      2 months         0.0             

   index intake_am_pm intake_time  
0  0      PM          2016-11-05   
1  1      AM          2016-09-28   
2  2      PM          2015-06-05   
3  3      PM          2015-05-29   
4  4      PM          2015-12-29

The next thing we observe is that there are still some attributes with datatype as '<b>object</b>' these are :
    <ul><li>Age upon Intake</li>
    <li>Breed_Intake</li>
    <li>Color_intake</li>
    <li>DateTime_Outcome</li>
    <li>Age upon Outcome</li></ul>
Further investigation into thses attributes will be carried out below


In [89]:
colums = ['Age upon Intake','Breed_Intake','Color_Intake','DateTime_Outcome','Age upon Outcome']

In [90]:
df[colums]

Age upon Intake                              Breed_Intake  \
0    5 months        Domestic Shorthair Mix                     
1    5 years         Pit Bull                                   
2    1 year          Domestic Shorthair Mix                     
3    4 weeks         Bat                                        
4    1 month         Labrador Retriever Mix                     
5    3 years         Bruss Griffon Mix                          
6    10 years        Pit Bull Mix                               
7    1 year          Miniature Poodle Mix                       
8    1 month         Border Terrier Mix                         
9    4 weeks         Domestic Shorthair Mix                     
10   2 years         Australian Kelpie Mix                      
11   11 months       Chihuahua Shorthair Mix                    
12   2 years         Domestic Medium Hair Mix                   
13   2 weeks         Domestic Medium Hair Mix                   
14   2 months        Pit Bull Mix                               
15   9 months        German Shepherd                            
16   8 months        German Shepherd Mix                        
17   1 year          Pit Bull Mix                               
18   1 year          Pomeranian/Chihuahua Longhair              
19   2 months        Labrador Retriever/Siberian Husky          
20   1 month         Domestic Shorthair Mix                     
21   9 months        Australian Cattle Dog Mix                  
22   2 years         Dachshund                                  
23   1 year          Chihuahua Shorthair Mix                    
24   1 month         Belgian Malinois Mix                       
25   1 year          Border Terrier Mix                         
26   7 years         Rottweiler Mix                             
27   5 years         Shih Tzu Mix                               
28   5 years         Domestic Shorthair Mix                     
29   3 years         Siberian Husky/American Pit Bull Terrier   
30   5 months        Pointer Mix                                
31   5 months        Rat Terrier Mix                            
32   6 years         Pit Bull Mix                               
33   4 weeks         Domestic Medium Hair Mix                   
34   1 month         Labrador Retriever/German Shepherd         
35   1 year          Domestic Shorthair Mix                     
36   6 years         Pit Bull Mix                               
37   2 years         Domestic Shorthair Mix                     
38   2 days          Domestic Shorthair Mix                     
39   5 months        Pointer Mix                                
40   3 years         Chihuahua Shorthair Mix                    
41   7 years         Domestic Shorthair Mix                     
42   1 year          Chihuahua Shorthair Mix                    
43   2 years         Bat Mix                                    
44   1 month         Domestic Shorthair                         
45   2 months        Basset Hound/Chinese Sharpei               
46   6 months        Domestic Shorthair Mix                     
47   3 weeks         Pit Bull Mix                               
48   2 months        Australian Cattle Dog                      
49   1 year          Boxer Mix                                  
50   2 years         Raccoon Mix                                
51   2 years         Shih Tzu Mix                               
52   5 years         Miniature Pinscher Mix                     
53   1 year          Great Pyrenees Mix                         
54   4 years         Yorkshire Terrier/Miniature Poodle         
55   10 years        German Shepherd Mix                        
56   2 months        Domestic Shorthair Mix                     
57   3 months        Russian Blue Mix                           
58   1 year          Labrador Retriever Mix                     
59   3 weeks         Domestic Shorthair                         
60   2 years         Mastiff Mix       

Looking at these attributes a number of problems are immediately apparent.
<ul>
    <li>The features 'Age upon Intake' and and 'Age upon Outcome' both express time period however non of them follow a common unit. Corrective measures such as converting to a single unit will be needed to make it viable data</li>
    <li> The attribute 'Breed_Intake' seems to have species and animal types mixed example there is both dog breeds and animal type like bats , chicken etc in same table which makes it difficult to seggregate. Another issue is the number of varieties of varieous species. Example in case of dogs. This posses a problem on how we can segregate species to animals , and wherether we can use this attribute for analysis</li>
    <li> Similar to the above attributes 'Color intake' attribute seems to have a high granularity and also seems to have multiple values for some animals as many animals have multiple colors; How to segregate or extract information from this is also need to be investigated</li>
    <li>The final attribute 'DateTime_Outcome' attribute indicates when the animal came out from shelter. However its my observation that this attribute gives same information that can be derived from attributes 'Age upon Intake' and 'Age upon Outcome'. Therfore this is a redundant attribute hence will be dropped</li>

Dropping DateTime_outcome from table 

In [91]:
df.drop(columns=['DateTime_Outcome'],inplace=True,axis=1)

Investigating the attribute 'Breed_intake'

In [92]:
df['Breed_Intake'].nunique()

219

There seems to 219 unique breeds in our dataset

In [93]:
temp = df['Breed_Intake'].value_counts()

In [94]:
display(pd.DataFrame(temp))

Breed_Intake
Domestic Shorthair Mix                    260         
Pit Bull Mix                              65          
Chihuahua Shorthair Mix                   59          
Labrador Retriever Mix                    50          
Domestic Shorthair                        38          
Domestic Medium Hair Mix                  35          
German Shepherd Mix                       24          
Domestic Longhair Mix                     18          
Bat Mix                                   18          
Australian Cattle Dog Mix                 16          
Bat                                       15          
Miniature Poodle Mix                      13          
Dachshund Mix                             11          
Siamese Mix                               10          
Pointer Mix                               8           
Raccoon Mix                               8           
Jack Russell Terrier Mix                  8           
Pit Bull                                  7           
Catahoula Mix                             7           
Rat Terrier Mix                           7           
Boxer Mix                                 7           
Cairn Terrier Mix                         6           
Great Pyrenees Mix                        5           
Raccoon                                   5           
Miniature Schnauzer Mix                   5           
German Shepherd                           5           
Australian Shepherd Mix                   5           
Chihuahua Shorthair                       5           
Border Collie Mix                         5           
Shih Tzu Mix                              4           
Labrador Retriever/German Shepherd        4           
Rottweiler Mix                            4           
Yorkshire Terrier Mix                     4           
Guinea Pig                                4           
Domestic Medium Hair                      4           
American Bulldog Mix                      4           
American Staffordshire Terrier Mix        4           
Plott Hound Mix                           4           
Mastiff Mix                               4           
Chihuahua Longhair Mix                    4           
German Shepherd/Labrador Retriever        4           
Staffordshire Mix                         3           
Dachshund                                 3           
Labrador Retriever/Pit Bull               3           
Siberian Husky Mix                        3           
Basset Hound Mix                          3           
Labrador Retriever                        3           
Black Mouth Cur Mix                       3           
Beagle Mix                                3           
Lhasa Apso Mix                            3           
Standard Poodle Mix                       3           
Queensland Heeler Mix                     3           
Belgian Malinois Mix                      2           
Doberman Pinsch                           2           
Ferret                                    2           
Border Terrier Mix                        2           
Dachshund/Chihuahua Shorthair             2           
Beagle/Chihuahua Shorthair                2           
Chihuahua Longhair                        2           
Himalayan Mix                             2           
Whippet Mix                               2           
Domestic Longhair                         2           
Bichon Frise Mix                          2           
Australian Kelpie Mix                     2           
Angora Mix                                2           
Bruss Griffon Mix                         2           
Miniature Schnauzer                       2           
Cairn Terrier                             2           
Siberian Husky                            2           
Miniature Pinscher Mix                    2           
American Pit Bull Terrier Mix             2           
English Shepherd Mix                      2           
Chinese Sharpei Mix        

Looking at the above table it seems that the attibute can provide some usefull insights. It seems that the most of animals lie in first 5 categories . Hence during the analysis we can bucket first 5 in to categories and rest in to another bucket called others. 

In [95]:
columns_with_high_count = df['Breed_Intake'].value_counts() > 10
x = df['Breed_Intake'].unique()
columns_with_high_count = x[columns_with_high_count]


Species which have occurance of more than 10 times. There are 14 such species and we will classify all other species as others as the occurance is really rare and will not be able to get any insights through them

In [96]:
display(pd.DataFrame(columns_with_high_count))

0
0   Domestic Shorthair Mix  
1   Pit Bull                
2   Bat                     
3   Labrador Retriever Mix  
4   Bruss Griffon Mix       
5   Pit Bull Mix            
6   Miniature Poodle Mix    
7   Border Terrier Mix      
8   Australian Kelpie Mix   
9   Chihuahua Shorthair Mix 
10  Domestic Medium Hair Mix
11  German Shepherd         
12  German Shepherd Mix

Here we take the 'Breed_intake' attribute then make all the uncommon species to type others

In [97]:
for x in range (1000):
    if df['Breed_Intake'][x] not in columns_with_high_count:
        df['Breed_Intake'][x] = "others"

/opt/anaconda3/envs/data_analytics/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


Investigating the attributes 'Age upon Intake' and 'Age upon Outtake'<br>
The attributes mentioned here define age however, the units are not standard , we need to make all the data into a single unit. One thing we will do is convert all data to type month. And also convert minimum time to 1 month so anything like 3 weeks will be converted to 1 month. The reason being we want to bin all the data for better visulaization and understandings. Also we see units year and years, month and months, week and weeks. This will also be removed. Also I will convert it to integer type

In [98]:
f =  df['Age upon Intake'][3].split()

In [99]:
for x in range(1000):
    temp = df['Age upon Intake'][x].split()
    if(temp[1] == 'week' or temp[1] == 'weeks'):
        df['Age upon Intake'][x] = 1
    elif(temp[1] == 'month'):
        df['Age upon Intake'][x] = 1
    elif(temp[1] == 'year'):
        df['Age upon Intake'][x] = 1
    elif(temp[1] == 'years'):
        df['Age upon Intake'][x] = (12*int(temp[0]))
    elif(temp[1] == 'months'):
        df['Age upon Intake'][x] = (int(temp[0]))
    elif(temp[1] == 'day'):
        df['Age upon Intake'][x] = 0
    elif(temp[1] == 'NaN'):
        df['Age upon Intake'][x] = 0

/opt/anaconda3/envs/data_analytics/lib/python3.7/site-packages/ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if sys.path[0] == '':
/opt/anaconda3/envs/data_analytics/lib/python3.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # Remove the CWD from sys.path while we load stuff.
/opt/anaconda3/envs/data_analytics/lib/python3.7/site-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.ht

Here we have converted all the units to months for attibute 'Age upon Intake'

In [100]:
df['Age upon Intake']

0      5     
1      60    
2      1     
3      1     
4      1     
5      36    
6      120   
7      1     
8      1     
9      1     
10     24    
11     11    
12     24    
13     1     
14     2     
15     9     
16     8     
17     1     
18     1     
19     2     
20     1     
21     9     
22     24    
23     1     
24     1     
25     1     
26     84    
27     60    
28     60    
29     36    
30     5     
31     5     
32     72    
33     1     
34     1     
35     1     
36     72    
37     24    
38     2 days
39     5     
40     36    
41     84    
42     1     
43     24    
44     1     
45     2     
46     6     
47     1     
48     2     
49     1     
50     24    
51     24    
52     60    
53     1     
54     48    
55     120   
56     2     
57     3     
58     1     
59     1     
60     24    
61     3     
62     3 days
63     156   
64     96    
65     1     
66     6     
67     1     
68     84    
69     24    
70     24    
71    

We do the same process for attribute 'Age upon Outake'

In [ ]:
# We have one element with value NaN which we will convert to 0 
df['Age upon Outcome'].fillna(0)

In [101]:


for x in range(1000):
    temp = df['Age upon Outcome'][x].split()
    if(temp[1] == 'week' or temp[1] == 'weeks'):
        df['Age upon Outcome'][x] = 1
    elif(temp[1] == 'month'):
        df['Age upon Outcome'][x] = 1
    elif(temp[1] == 'year'):
        df['Age upon Outcome'][x] = 1
    elif(temp[1] == 'years'):
        df['Age upon Outcome'][x] = (12*int(temp[0]))
    elif(temp[1] == 'months'):
        df['Age upon Outcome'][x] = (int(temp[0]))
    elif(temp[1] == 'day'):
        df['Age upon Outcome'][x] = 0
    elif(temp[1] == 'days'):
        df['Age upon Outcome'][x] = 0
    elif(temp[1] == 'NaN'):
        df['Age upon Outcome'][x] = 0

/opt/anaconda3/envs/data_analytics/lib/python3.7/site-packages/ipykernel_launcher.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  from ipykernel import kernelapp as app
/opt/anaconda3/envs/data_analytics/lib/python3.7/site-packages/ipykernel_launcher.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  del sys.path[0]
/opt/anaconda3/envs/data_analytics/lib/python3.7/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-vie

AttributeError: 'float' object has no attribute 'split'

In [102]:
df['Age upon Outcome']

0      5        
1      60       
2      1        
3      1        
4      2        
5      36       
6      120      
7      1        
8      2        
9      1        
10     24       
11     11       
12     24       
13     1        
14     2        
15     9        
16     8        
17     10       
18     1        
19     2        
20     9        
21     10       
22     24       
23     1        
24     2        
25     1        
26     84       
27     60       
28     60       
29     36       
30     6        
31     5        
32     72       
33     2        
34     1        
35     1        
36     72       
37     24       
38     0        
39     7        
40     36       
41     84       
42     1        
43     24       
44     3        
45     3        
46     6        
47     1        
48     2        
49     1        
50     24       
51     24       
52     60       
53     1        
54     48       
55     120      
56     3        
57     3        
58     1      

which month most intake?


<h4> Steps completed uptill now</h4>
<ul><li>Checked for constant value attributes</li>
    <li>Checked and removed duplicate columns</li>
    <li>Converted all attributes to appropriate data types</li>
    <li>Checked for null values and see if some form of imputations can be carried out</li>
    <li>Removed attibutes which didn't provide any useful information</li>
    <li> Checked for attributes with anomalous cardinalities</li>
    <li>Extracted new usefull features from existing attributed</li>
    <li>Reduced the granularity of some attributes</li>